### ISSUE TITLE
Problem with LimitStencilTableFactory on latest dev branch commit

### ISSUE THREAD

**Issue by oracle3001** [PixarAnimationStudios/OpenSubdiv#1028]
The following block of code works without error on the release branch (and on a previous out of date build of the dev branch, which was a few commits behind *), but when I switch to a build of the latest commit dev branch, it returns a null g_controlStencils object (because it has internally errored). 

The refinement being used is a Uniform Catmull Clark.

This on VS2017 on Windows 10 and occurs no matter what the original input mesh is.

*I am afraid I can't tell you how far back on the dev branch I was, but not that many commits. 

### Edit - I have rolled back to 1st Nov commit, and it works fine. So something since then is causing this issue.

```
std::vector<LimitStencilTableFactory::LocationArray> locs(_nFaces);

	float *u = new float[_nFaces], *uPtr = u,
		*v = new float[_nFaces], *vPtr = v;

	for (auto f : range_n(static_cast<int>(_nFaces)))
	{
		LimitStencilTableFactory::LocationArray& larray = locs[f];

		larray.ptexIdx = f;
		larray.numLocations = 1;

		*uPtr = static_cast<float>(0.5);
		*vPtr = static_cast<float>(0.5);

		larray.s = uPtr;
		larray.t = vPtr;

		++uPtr;
		++vPtr;
	}


	StencilTableFactory::Options stencilTableOptions;
	stencilTableOptions.generateIntermediateLevels = true;
	stencilTableOptions.generateControlVerts = true;
	stencilTableOptions.generateOffsets = true;

	auto cvstencils = StencilTableFactory::Create(*_refiner, stencilTableOptions);

	if (StencilTable const* localPointStencilTable =
		_patchTable->GetLocalPointStencilTable())
	{
		StencilTable const* table =
			StencilTableFactory::AppendLocalPointStencilTable(
				*_refiner, cvstencils, localPointStencilTable);
		delete cvstencils;
		cvstencils = table;
	}

	LimitStencilTableFactory::Options st_options;
	st_options.generate2ndDerivatives = false;
	auto g_controlStencils = LimitStencilTableFactory::Create(*_refiner, locs, cvstencils, _patchTable, st_options);
```

If I replace the last line with,

`auto g_controlStencils = LimitStencilTableFactory::Create(*_refiner, locs);`

i.e let the LimitStencilTableFactory build the cvstencils and patchTable, I get an assert error in stencilTableFactory.cpp at line 477

`assert(cvstencils->GetNumStencils() == nRefinedStencils);`

**Comment by barfowl** [comment]
Thanks for the details -- and for the update that I just noticed.

I was going to say that we made additions to the LimitStencilTableFactory back in #1012 and that it was likely the cause.  But that was merged on Oct 29 -- just prior to the November commits you're now using that are working fine.

This is working fine with the compilers I have access to and the versions of VS we've been testing so far, so I'm working in the dark a bit.  If you do feel inclined to pinpointing the commit at which the problem occurs, that would be really helpful.  Meanwhile, knowing #1012 works and that the problem was introduced later gives me something more to go on.  Thanks

**Comment by oracle3001** [comment]
I have just checked and the next verified pull #1013 (5th November) is where it appears to cause issues.

**Comment by barfowl** [comment]
Excellent, that should be a big help, thanks.  (Odd though -- that commit was all about patches and the problems here occur in creating a StencilTable from the adaptive refinement...)

**Comment by barfowl** [comment]
I was looking at #1029 -- which occurs for a specific version of VS2017 -- and was wondering if the same is true here.  Is this at least occurring with the same VS2017 version that causes #1029's failure?  Thanks

**Comment by oracle3001** [comment]
As far as I can tell, #1029 occurs only when compiling specifically with VS2017.

The above issue definitely occurs when I compile with either VS2017 or LLVM / Clang.

**Comment by jtran56** [comment]
Filed as internal issue #OSD-249.

**Comment by oracle3001** [comment]
Just to add, I have played around a little bit with this with different meshes etc and it is definitely only an issue when using UNIFORM subdivision. 

If I leave my code exactly the same except change the subdivision type to ADAPTIVE it doesn't crash.

**Comment by barfowl** [comment]
I've identified a regression for the Uniform case that was introduced in #1012, and I don't think the topology option is the issue here.  I would be more confident that it address your problem if it weren't for the fact that you mentioned it doesn't start happening until the commit for #1013 was merged.

There is another oversight for the Uniform case that I need to address relative to #1012 so I won't have a pull request ready for another day or two.

In the meantime, you might want to confirm if the issue really starts happening with #1012.  That is when the LimitStencilTableFactory was modified to support face-varying and the Uniform case clearly suffered from a lack of testing coverage.  I'm happy to push a personal branch with a fix to try and confirm if you want while I work on the related issues for the upcoming pull request.

Thanks

**Comment by barfowl** [comment]
I'm fairly confident #1031 (merged earlier today) will address this problem.  Please try it out when you get the chance and close this if it does so.  Thanks

**Comment by oracle3001** [comment]
Ok, thank you. It will be a few days before I will get chance to look at this, but I will test it next week.

**Comment by barfowl** [comment]
Just checking in on this issue to see if any problems still persist, thanks.

**Comment by oracle3001** [comment]
Sorry for not replying. As far as I know, no problems now.

**Comment by barfowl** [comment]
Closing -- should be fixed as of #1031.
